tempodb/encoding/iterator_multiblock.go

```diff
@@ -0,0 +1,194 @@
+package encoding
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/grafana/tempo/tempodb/encoding/common"
+)
+
+type multiblockIterator struct {
+	combiner     common.ObjectCombiner
+	bookmarks    []*bookmark
+	dataEncoding string
+	ctx          context.Context
+	resultsCh    chan iteratorResult
+	quitCh       chan bool
+
+	errMtx sync.RWMutex
+	err    error
+}
+
+type iteratorResult struct {
+	ID   common.ID
+	Body []byte
+}
+
+// NewMultiblockIterator Creates a new multiblock iterator. Iterates concurrently in a separate goroutine and results are buffered.
+// Traces are deduped and combined using the object combiner.
+func NewMultiblockIterator(ctx context.Context, inputs []Iterator, bufferSize int, combiner common.ObjectCombiner, dataEncoding string) Iterator {
+	i := multiblockIterator{
+		ctx:          ctx,
+		combiner:     combiner,
+		dataEncoding: dataEncoding,
+		resultsCh:    make(chan iteratorResult, bufferSize),
+		quitCh:       make(chan bool, 1),
+	}
+
+	for _, iter := range inputs {
+		i.bookmarks = append(i.bookmarks, newBookmark(iter))
+	}
+
+	go i.iterate()
+
+	return &i
+}
+
+func (i *multiblockIterator) Close() {
+	select {
+	case i.quitCh <- true:
+	default:
+		return
+	}
+}
+
+func (i *multiblockIterator) Next(ctx context.Context) (common.ID, []byte, error) {
+	if err := i.getErr(); err != nil {
+		return nil, nil, err
+	}
+
+	select {
+	case <-i.ctx.Done():
```

Comment:
i believe this should use the passed in context and we shouldn't store a context in the struct.